File: backend/routers/moderation.py
import re
import base64
from fastapi import APIRouter
from pydantic import BaseModel

router = APIRouter()

SEXUAL_PATTERNS = [
    b'g+e/gA==', b'5aaT5aWz', b'5LiK5bqK', b'57qm54iu', b'5byA5oi/',
    b'6KOk5L2T', b'6Imy5oOF', b'6buE54mH', b'5YGa54ix', b'6IOW5LiL',
    b'6Imy5q+N', b'c2V4', b'bnVkZQ==', b'cG9ybg==',
]

POLITICAL_PATTERNS = [
    b'5YWt5Zub', b'5q+b5rO95Lic', b'5paH6Z2p', b'5rOV6L2u5Yqf',
    b'5YWx5Lqn5YWa', b'5rSq56eR', b'5Y+N5YWx',
]

VIOLENT_PATTERNS = [
    b'5p2A5Lq6', b'5pq06Kqo', b'54K45by5', b'c3VpY2lkZQ==',
    b'6Ieq5p2A',
]

SPAM_PATTERNS = [
    b'5Yqg5b6u5L+h', b'V2VDaGF0', b'5LqM57u0', b'5YWN6LS5',
    b'6LWa6ZKx',
]


def _decode_patterns(encoded: list[bytes]) -> list[str]:
    result = []
    for b in encoded:
        try:
            result.append(base64.b64decode(b).decode("utf-8"))
        except Exception:
            pass
    return result


DECODED = {
    "sexual": _decode_patterns(SEXUAL_PATTERNS),
    "political": _decode_patterns(POLITICAL_PATTERNS),
    "violent": _decode_patterns(VIOLENT_PATTERNS),
    "spam": _decode_patterns(SPAM_PATTERNS),
}

REASON_MESSAGES = {
    "sexual": "消息包含不当内容，请注意文明用语",
    "political": "消息包含敏感内容，请修改后重试",
    "violent": "消息包含暴力相关内容，请修改",
    "spam": "消息疑似广告或诈骗信息，请确认",
}


def _normalize(text: str) -> str:
    return re.sub(r"[\s​‌‍﻿，。！？、；：""''【】（）《》…—·\-_.!?,;:'\"()\[\]{}]", "", text.lower())


def check_content(text: str) -> dict:
    if not text or not text.strip():
        return {"safe": True}

    norm = _normalize(text)
    for reason, patterns in DECODED.items():
        for p in patterns:
            np = _normalize(p)
            if np and np in norm:
                return {"safe": False, "reason": reason, "message": REASON_MESSAGES[reason]}

    return {"safe": True}


class ModerationRequest(BaseModel):
    content: str


@router.post("/check")
async def check(req: ModerationRequest):
    return check_content(req.content)

File: frontend/src/services/moderation.ts
export interface ModerationResult {
  safe: boolean
  reason?: 'sexual' | 'political' | 'violent' | 'spam' | 'other'
  message?: string
}

// Base64-encoded pattern groups to avoid plain-text sensitive words in source
// Each entry: base64(word) — decoded at runtime for matching
const SEXUAL_PATTERNS = [
  'g+e/gA==', '5aaT5aWz', '5LiK5bqK', '57qm54iu', '5byA5oi/', '6KOk5L2T',
  '6Imy5oOF', '6buE54mH', '5YGa54ix', '6IOW5LiL', '4piF4piF4piF',
  '6Imy5q+N', 'c2V4', 'bnVkZQ==', 'cG9ybg==',
].map(b => atob(b))

const POLITICAL_PATTERNS = [
  '+mUU4pa3', '5YWt5Zub', '5q+b5rO95Lic', '5paH6Z2p', '5rOV6L2u5Yqf',
  '5Lmg6L+R5bmz', '6YKT5bCP5bmz', '5YWx5Lqn5YWa', '5rSq56eR', '5Y+N5YWx',
  '57uP57u/', '5rKm6Zm3', '5rOV6L2u', '5YWo6IO95',
].map(b => { try { return atob(b) } catch { return '' } }).filter(Boolean)

const VIOLENT_PATTERNS = [
  '5p2A5Lq6', '5pq06Kqo', '54K45by5', '5q275LiA5Liq', 'c3VpY2lkZQ==',
  '6Ieq5p2A', '5omp5Yiw', '5by55q2s',
].map(b => { try { return atob(b) } catch { return '' } }).filter(Boolean)

const SPAM_PATTERNS = [
  '5Yqg5b6u5L+h', 'V2VDaGF0', '5LqM57u0', '5YWN6LS5',
  '6LWa6ZKx', '5Yiw5omL', '6L2s6LSm',
].map(b => { try { return atob(b) } catch { return '' } }).filter(Boolean)

const REASON_MESSAGES: Record<string, string> = {
  sexual: '消息包含不当内容，请注意文明用语',
  political: '消息包含敏感内容，请修改后重试',
  violent: '消息包含暴力相关内容，请修改',
  spam: '消息疑似广告或诈骗信息，请确认',
}

function normalize(text: string): string {
  return text
    .toLowerCase()
    .replace(/[\s​‌‍﻿]/g, '')
    .replace(/[，。！？、；：""''【】（）《》…—·\-_.!?,;:'"()\[\]{}]/g, '')
}

function matchPatterns(text: string, patterns: string[]): boolean {
  const norm = normalize(text)
  return patterns.some(p => {
    const np = normalize(p)
    return np.length > 0 && norm.includes(np)
  })
}

export function moderateContent(text: string): ModerationResult {
  if (!text || text.trim().length === 0) return { safe: true }

  if (matchPatterns(text, SEXUAL_PATTERNS)) {
    return { safe: false, reason: 'sexual', message: REASON_MESSAGES.sexual }
  }
  if (matchPatterns(text, POLITICAL_PATTERNS)) {
    return { safe: false, reason: 'political', message: REASON_MESSAGES.political }
  }
  if (matchPatterns(text, VIOLENT_PATTERNS)) {
    return { safe: false, reason: 'violent', message: REASON_MESSAGES.violent }
  }
  if (matchPatterns(text, SPAM_PATTERNS)) {
    return { safe: false, reason: 'spam', message: REASON_MESSAGES.spam }
  }

  return { safe: true }
}

const API_BASE = import.meta.env.VITE_API_URL || 'http://localhost:8000'

export async function moderateContentAsync(text: string): Promise<ModerationResult> {
  try {
    const res = await fetch(`${API_BASE}/api/moderation/check`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content: text }),
    })
    if (!res.ok) return { safe: true }
    return await res.json()
  } catch {
    return { safe: true }
  }
}
